Protected build parameters

jlaitine opened the pull request:
This adds another library for parameters interface for protected build user space. The protected build uses the same interface for kernel side, as what is used in flat build. The user space library is just a collection of interface functions, which perform IOCTL calls to the parameters libary in kernel.

This all makes parameters interface more like a kernel side driver in protected build.

The flat build parameters library is not changed at all.

Originally I had an implementation, where the whole parameters library was just duplicated in both kernel and user sides, but it didn't really work properly. The locking of parameter storage obviously failed, and also the parameter changes from user side were not propagated properly to kernel side when they were changed in user side; except after next reboot. After struggling with that implementation for a while, I decided to do this "driver" like approach instead, which is a tad heavier in CPU load, but works the same as in flat build (correctly).

jlaitine commented on src/lib/parameters/CMakeLists.txt:
```diff
@@ -169,7 +178,7 @@ if (NOT "${PX4_BOARD}" MATCHES "px4_io")
 	target_link_libraries(parameters PRIVATE perf tinybson px4_platform)
 
-	target_compile_definitions(parameters PRIVATE -DMODULE_NAME="parameters")
+	target_compile_definitions(parameters PRIVATE -DMODULE_NAME="parameters" -D__KERNEL__)
```
FYI: the __KERNEL__ flag here is needed only for nuttx includes. Without this, the nuttx work queues are allocated wrongly. The flag is not used anywhere in posix build, and has not effect in flat build.

jlaitine commented:
@dagar, @bkueng I wonder if you had time to check this; there is not much left for "protected build" support

dagar commented:
@jlaitine at this point can you bring in a protected board that we can add to the builds and run on the test rack?

jlaitine commented:
> @jlaitine at this point can you bring in a protected board that we can add to the builds and run on the test rack?

Yes, on top of this I can add the config for fmu-v5, which I habe have got ready. #16931 actually has everything. The crypto if additions there are not needed to run the thing, the crypto support can be left out initially.

dagar commented:
To prevent regressions from slipping in I think it would really help to have it regularly building and running.

jlaitine commented:
I will collect.the minimum set of patches to get it building and running on pixhawk 4 tomorrow

jlaitine commented:
> To prevent regressions from slipping in I think it would really help to have it regularly building and running.

I updated  #16931 to contain the minimum to enable px4_fmu-v5_protected target for pixhawk 4 hardware. If we want to have the protected target building and running right away, that PR should be taken in then, and this one can be discarded. That PR has also flown on holybro x500 frame.

I dropped some crypto interface support and support for reading kernel-side perf counters from that PR, they can be added later, it works without those. Also there is still some patches missing from NuttX w.r.t user-side workqueues. They come in automatically if nuttx is updated to newer version, or I can backport those later. The missing nuttx patches affect mainly cdc-acm detection.

If you think that that PR is too large, then we need to continue taking in smaller pieces still before the target can be enabled...

bkueng commented on src/lib/parameters/usr_parameters_if.cpp:
```diff
@@ -0,0 +1,220 @@
+{
+	paramiocexport_t data = {filename, PX4_ERROR};
+
+	if (filter) { PX4_ERR("ERROR: filter not supported in userside blob"); }
```
Note that this is used for factory calibration: https://github.com/PX4/PX4-Autopilot/blob/master/src%2Fmodules%2Fcommander%2Ffactory_calibration_storage.cpp#L83 (not immediately relevant)

jlaitine replied:
Yes; if the fileter is needed in user side, I need to come back to this. I admit that I just took a shortcut here and added the ERROR to remind if this pops up...

bkueng commented on src/lib/parameters/parameters.cpp:
```diff
@@ -69,6 +69,14 @@ using namespace time_literals;
 #include <uORB/Subscription.hpp>
 
+/* Include functions common to user and kernel sides */
+#include "parameters_common.cpp"
```
I prefer having this as a header, you can make most of the methods `static inline`.

jlaitine replied:
making these static inline is actually not that simple. The generated px4_parameters.hpp can be included just once, otherwise the tables will get multiplied increasing the code size. And just re-naming the file to .h ends up style checks failing in CI. Need to think about this, what might be the best alternative way to avoid having duplicate code.

bkueng replied:
Ok not a big deal, it's just a bit of an anti-pattern.

bkueng reviewed (commented):
Looks good, I'm fine with getting this in as it is.

jlaitine commented:
@bkueng if this is ok, please merge; I can then rebase https://github.com/PX4/PX4-Autopilot/pull/16931 to add the rest of the initializations and one board config (for pixhawk 4). So we can finally get a building&running protected build configuration in CI
